### PR TITLE
Ruby: rack - model more responses and app types

### PR DESCRIPTION
**Pull request by alexrford**
This is a follow-up to https://github.com/github/codeql/pull/13289.

Rack responses can be constructed using calling `finish` on a `Rack::Response` instance. This may be more convenient in some cases (e.g. default values are set when calling `Rack::Response.new` if not specified). This PR adds modelling to consider these calls to `finish` as creating a rack response.

This PR also extends recognition of rack apps to include:
 1. Cases where the `call(env)` method is a class method rather than an instance method
 2. "Anonymous" applications, e.g. those defined as lambdas or procs
 
In the case of anonymous applications we limit candidates to those which flow to arg0 of a `run` call, as considering all lambdas/procs can be prohibitively expensive.

### CODE REVIEW

File: ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowPublic.qll [hunk 1]
```diff
@@ -1010,6 +1010,23 @@ class ModuleNode instanceof Module {
...
 
+  /**
+   * Gets the singleton method named `name` available in this module, including methods inherited from ancestors.
+   *
+   * Overridden methods are not included.
+   */
+  MethodNode getSingletonMethod(string name) {
+    result.asCallableAstNode() = super.getAnOwnSingletonMethod() and
+    result.getMethodName() = name
+  }
```
[the remark this answers is not in the document]
Comment (alexrford) [reply ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowPublic.qll:1021]: 🤦 thanks, yes the body doesn't match. We don't need this predicate if we drop the distinction between class apps and anonymous apps, so I'll just remove this predicate.

File: ruby/ql/lib/codeql/ruby/frameworks/rack/internal/App.qll [hunk 2]
```diff
@@ -2,24 +2,56 @@
...
 private import Response::Private as RP
 
-/** A method node for a method named `call`. */
-private class CallMethodNode extends DataFlow::MethodNode {
-  CallMethodNode() { this.getMethodName() = "call" }
+/**
+ * A callable node that takes a single argument and, if it has a method name,
+ * is called "call".
+ */
+private class PotentialCallNode extends DataFlow::CallableNode {
```
[the remark this answers is not in the document]
Comment (alexrford) [reply ruby/ql/lib/codeql/ruby/frameworks/rack/internal/App.qll:15]: Renamed this as `PotentialRequestHandler` and the version where we track the response type as `RequestHandler`.

File: ruby/ql/lib/codeql/ruby/frameworks/rack/internal/App.qll [hunk 3]
```diff
@@ -50,4 +83,65 @@ module App {
     /** Gets the response returned from a request to this application. */
     RP::PotentialResponseNode getResponse() { result = resp }
   }
+
+  private newtype TApp =
+    TClassApp(DataFlow::ClassNode cn, CallNode call) or
+    TAnonymousApp(CallNode call)
```
[the remark this answers is not in the document]
Comment (alexrford) [reply ruby/ql/lib/codeql/ruby/frameworks/rack/internal/App.qll:89]: The main purpose here was to maintain similar behaviour as before where we considered the class containing the `call` method as the app if such a class existed. In practice, I don't think it matters? As you mention, we can figure this out from the `RequestHandler` if it ends up being relevant.

On that note I've moved the `RequestHandler` itself to replace the `RackApplication` class - this is the `DataFlow::CallableNode` corresponding to the method/proc/lambda that takes the `env` and returns a rack compliant response.

**Comment by alexrford** [comment]
@asgerf this is updated - as we briefly discussed, recognition of request handlers can likely be improved using your API graphs overhaul. The overall diff for this PR is simpler than before. I'm interested in if the publicly exposed API for `Rack::App::RequestHandler` looks reasonable, as the precise implementation is liable to change soon anyway.

File: ruby/ql/lib/codeql/ruby/frameworks/rack/internal/App.qll [hunk 4]
```diff
@@ -2,24 +2,44 @@
...
+  PotentialRequestHandler() {
+    this.getNumberOfParameters() = 1 and
+    (
+      this.(DataFlow::MethodNode).getMethodName() = "call"
+      or
+      not this instanceof DataFlow::MethodNode and
+      exists(DataFlow::CallNode cn | cn.getMethodName() = "run" |
+        this.(DataFlow::LocalSourceNode).flowsTo(cn.getArgument(0))
+        or
+        // TODO: `Proc.new` should automatically propagate flow from its block argument
```
Comment (asgerf) [inline ruby/ql/lib/codeql/ruby/frameworks/rack/internal/App.qll:25]: I've opened a PR here to fix this https://github.com/github/codeql/pull/13612

After that gets merged we should be able to write the entire following clause as:
```codeql
this = API::getTopLevelCall("run").getArgument(0).asCallable()
```

Comment (alexrford) [reply ruby/ql/lib/codeql/ruby/frameworks/rack/internal/App.qll:25]: I'll drop the explicit proc handling once that PR is in.

Comment (asgerf) [reply ruby/ql/lib/codeql/ruby/frameworks/rack/internal/App.qll:25]: Since the PR is not able to be merged right now, let's leave it like this for now

**Comment by asgerf** [comment]
I've started another evaluation with meta-queries as I'd like a closer look at what sinks are found by this heuristic in practice. LGTM if the results look reasonable.

**Review by asgerf — APPROVED** [review]
LGTM - the evaluation was quiet

**Pull request merged**